File: Matlab_codes/3. Call price stock option/callOR.m
function callOR = callOR(eta1, eta2, la, p, sig, rr, bigS, bigK,  bigT, nStep)

    function zetaaOR
        zetaaOR = p*eta1 /(eta1 -1) + (1-p)*eta2/(eta2+1) -1;
    end

    function tempaa1OR
      tempaa1OR =  rr + sig*sig/2 - la*zetaaOR;
    end

    function tempaa2OR
      tempaa2OR = tempaa1OR - sig*sig;
    end

callOR = bigS *...
 cprob(tempaa1OR, eta1 - 1, eta2 + 1, la *(1+zetaaOR), p*eta1/((1+zetaaOR)*(eta1-1)), sig, Log(bigK/bigS),  bigT, nStep) -...
  bigK * Exp(- rr*bigT) *cprob(tempaa2OR, eta1, eta2, la, p, sig, Log(bigK/bigS),  bigT, nStep)
end

File: Matlab_codes/3. Call price stock option/callOR.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

import math

def callOR(eta1, eta2, la, p, sig, rr, bigS, bigK, bigT, nStep):

    def zetaaOR():
        return p * eta1 / (eta1 - 1) + (1 - p) * eta2 / (eta2 + 1) - 1

    def tempaa1OR():
        return rr + sig * sig / 2 - la * zetaaOR()

    def tempaa2OR():
        return tempaa1OR() - sig * sig

    def cprob(x, alpha, beta, lambda_val, p_val, sigma, log_KS, T, n):
        # Implement the cprob function according to your requirements
        # This is a placeholder, replace it with the actual implementation
        return 0

    callOR = bigS * cprob(tempaa1OR(), eta1 - 1, eta2 + 1, la * (1 + zetaaOR()), p * eta1 / ((1 + zetaaOR()) * (eta1 - 1)), sig,
                         math.log(bigK / bigS), bigT, nStep) - bigK * math.exp(-rr * bigT) * cprob(tempaa2OR(), eta1,
                                                                                                    eta2, la, p, sig,
                                                                                                    math.log(
                                                                                                        bigK / bigS),
                                                                                                    bigT, nStep)
    return callOR
